### ISSUE TITLE
Multi Task Learning

### ISSUE THREAD

**Issue by Kalkasas** [ppwwyyxx/tensorpack#29]
Does support for multi task learning exist? If not, which way would fit this architecture? I am talking about a setup where a new task is chosen for each minibatch pass.

Thank you for any help you can provide.

**Comment by ppwwyyxx** [comment]
If it is a cost function of several different cost, you just simply add them.
If each time the data is different, you can implement such DataFlow that generates different data for different tasks, as well as an indicator so that the model knows how to deal with them.

**Comment by Kalkasas** [comment]
Thank you for your fast reply.

Lets assume I have n tasks (n linear classifiers at the end of the network, the labels of al tasks have the same "type/dimension" but different meaning). At the moment I treat a task ID simply as an input value which flows through the net. I stored my n linear classifier weights in a 2-D Tensor (first dimension is task, second dimension is weights of the tasks).

The task is now simply used in tf.gather to pick the corresponding weights for the task. As I am using a momentum optimizer I do not know if some weights of an "idle" task could changed as well. Mulitple sources I found on the web suggest that multiple optimizers should be used.

Could I run into issues with my setup?

**Comment by ppwwyyxx** [comment]
You can simply use indexing `W[row]` or `W[:,column]` to select the weights, where `row` or `column` is an integer scalar tf.Tensor. The gradient would work well. `tf.gather` would also work. Only the weights for that task would change.

**Comment by laceyg** [comment]
I am also interested in implementing a multi-task learning model using tensorpack - similar to the "alternating training" example in https://jg8610.github.io/Multi-Task/.  In this example, there are 2 different datasets for 2 tasks, and we want to train a model which uses 1 shared layer, and 1 task-specific layer for each task.  While the example using plain tensorflow is clear, I am not sure what the best approach is using tensorpack.

If we generate two different DataFlow objects such that each generates different data for each task, how would you use the ```SyncMultiGPUTrainer``` with two different DataFlow objects?  Since this trainer uses the ```QueueInputTrainer``` is it possible to be enqueuing two different DataFlow objects?

Also, for the task specific layers, would it better to use separate layers with separate cost functions (as in the example) or a single layer with selectable weights? Is there an easy way to use an indicator to select the appropriate weights during training?

Has anyone done anything similar / have any advice?

**Comment by ppwwyyxx** [comment]
Above question answered in #35 .
